src/Confluent.Kafka/Consumer.cs

```diff
@@ -1,175 +1,346 @@
+
+
+        public WatermarkOffsets GetWatermarkOffsets(TopicPartition topicPartition)
+            => consumer.GetWatermarkOffsets(topicPartition);
+
+
+        public WatermarkOffsets QueryWatermarkOffsets(TopicPartition topicPartition, TimeSpan timeout)
+            => consumer.QueryWatermarkOffsets(topicPartition, timeout);
+
+        public WatermarkOffsets QueryWatermarkOffsets(TopicPartition topicPartition)
+            => consumer.QueryWatermarkOffsets(topicPartition);
+
+
+        public Metadata GetMetadata(bool allTopics, TimeSpan timeout)
+            => consumer.GetMetadata(allTopics, timeout);
+
+        public Metadata GetMetadata(bool allTopics)
+            => consumer.GetMetadata(allTopics);
+    }
+
+    public class Consumer : IDisposable
+    {
+        private CancellationTokenSource consumerCts = null;
+        private Task consumerTask = null;
+
+        private SafeKafkaHandle kafkaHandle;
+
+        private void ErrorCallback(IntPtr rk, ErrorCode err, string reason, IntPtr opaque)
+        {
+            // TODO: Is reason ever different from that returned by err2str?
+            //       If so, sort something else out here.
+            OnError?.Invoke(this, err);
+        }
+
+        private int StatsCallback(IntPtr rk, IntPtr json, UIntPtr json_len, IntPtr opaque)
+        {
+            OnStatistics?.Invoke(this, Util.Marshal.PtrToStringUTF8(json));
+            return 0; // instruct librdkafka to immediately free the json ptr.
+        }
 
-        // Rebalance callbacks
-        public event EventHandler<List<TopicPartitionOffset>> OnPartitionsAssigned;
-        public event EventHandler<List<TopicPartitionOffset>> OnPartitionsRevoked;
+        private void LogCallback(IntPtr rk, int level, string fac, string buf)
+        {
+            var name = Util.Marshal.PtrToStringUTF8(LibRdKafka.name(rk));
+
+            if (OnLog == null)
+            {
+                // A stderr logger is used by default if none is specified.
+                Loggers.ConsoleLogger(this, new LogMessage(name, level, fac, buf));
+                return;
+            }
+
+            OnLog?.Invoke(this, new LogMessage(name, level, fac, buf));
+
+        }
 
         // Explicitly keep reference to delegate so it stays alive
-        LibRdKafka.RebalanceCallback RebalanceDelegate;
-        void RebalanceCallback(IntPtr rk, ErrorCode err,
-                /* rd_kafka_topic_partition_list_t * */ IntPtr partitions,
-                IntPtr opaque)
+        private LibRdKafka.RebalanceDelegate rebalanceDelegate;
+        private void RebalanceCallback(
+            IntPtr rk,
+            ErrorCode err,
+            /* rd_kafka_topic_partition_list_t * */ IntPtr partitions,
+            IntPtr opaque)
         {
-            var partitionList = SafeKafkaHandle.GetTopicPartitionOffsetList(partitions);
+            var partitionList = SafeKafkaHandle.GetTopicPartitionOffsetErrorList(partitions).Select(p => p.TopicPartition).ToList();
```

Comment:
@mhowlett @edenhill 
I just came accross that. Can you explain why we don't retrieve the offset and pass Offset.Invalid instead? I must be missing something.